client/comms/wsconn.go

```diff
@@ -289,35 +340,43 @@ func (conn *wsConn) NextID() uint64 {
 // Connect connects the client and starts an auto-reconnect loop. Any error
 // encountered during the initial connection will be returned. The reconnect
 // loop will continue to try connecting, even if an error is returned. To
-// shutdown auto-reconnect, use close().
+// shutdown auto-reconnect, use Stop() or cancel the context.
 func (conn *wsConn) Connect(ctx context.Context) (error, *sync.WaitGroup) {
-	var wg sync.WaitGroup
-	wg.Add(1)
-
-	go conn.keepAlive(ctx)
+	var ctxInternal context.Context
+	ctxInternal, conn.cancel = context.WithCancel(ctx)
 
+	conn.wg.Add(1)
 	go func() {
-		<-ctx.Done()
-		conn.close()
-		wg.Done()
+		defer conn.wg.Done()
+		conn.keepAlive(ctxInternal)
 	}()
 
-	return conn.connect(), &wg
-}
-
-// Close terminates all websocket processes and closes the connection.
-func (conn *wsConn) close() {
-	conn.wsMtx.Lock()
-	defer conn.wsMtx.Unlock()
+	conn.wg.Add(1)
+	go func() {
+		defer conn.wg.Done()
+		<-ctxInternal.Done()
+		conn.setConnected(false)
+		if conn.ws != nil {
+			log.Debug("Sending close 1000 (normal) message.")
+			msg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "bye")
+			conn.ws.WriteControl(websocket.CloseMessage, msg,
+				time.Now().Add(writeWait))
+			conn.ws.Close()
+		}
+		close(conn.readCh) // signal to receivers that the wsConn is dead
+	}()
 
-	if conn.ws == nil {
-		return
+	err := conn.connect(ctxInternal)
+	if err != nil {
+		conn.queueReconnect()
 	}
```

Comment:
This breaks the expected behavior. One major purpose of `Connector` is that if an error is encountered in the initial connection, the error is returned and the `Connector` is expected to remain disconnected, letting the caller decide whether and when a reconnection is attempted. Only disconnects that occur after a connection is successfully negotiated will be eligible for auto-reconnect.

Reply:
I can remove the `queueReconnect`, but `connect` was previously doing auto-reconnect right off the bat.

Reply:
BTW, I strongly prefer the semantics you described.  Trying to reconnect repeatedly when it's failing on e.g. Dial because of a bad host/port is silly. It's changed that way now.